src/frequenz/sdk/timeseries/_ringbuffer.py

```diff
@@ -0,0 +1,430 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency."""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Any, Generic, List, Sequence, TypeVar
+
+import numpy as np
+
+T = TypeVar("T")
+
+Container = TypeVar("Container", list, np.ndarray)
+
+
+@dataclass
+class Gap:
+    """A gap defines the range for which we haven't received values yet."""
+
+    start: datetime
+    """Start of the range, inclusive."""
+    end: datetime
+    """End of the range, exclusive."""
+
+    def contains(self, timestamp: datetime):
+        """Check if a given timestamp is inside this  gap.
+
+        Args:
+            timestamp: Timestamp to check.
+
+        Returns:
+            True if the timestamp is in the gap.
+        """
+        if self.start <= timestamp < self.end:
+            return True
+
+        return False
+
+
+class OrderedRingBuffer(Generic[T, Container]):
+    """Time aware ringbuffer that keeps its entries sorted by time."""
+
+    def __init__(
+        self,
+        buffer: Container,
+        sampling_period: timedelta,
+        time_index_alignment: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """Initialize the time aware ringbuffer.
+
+        Args:
+            buffer: Instance of a buffer container to use internally.
+            sampling_period: Timedelta of the desired resampling period.
+            time_index_alignment: Arbitary point in time used to align
+                timestamped data with the index position in the buffer.
+                Used to make the data stored in the buffer align with the
+                beginning and end of the buffer borders.
+
+                For example, if the `time_index_alignment` is set to
+                "0001-01-01 12:00:00", and the `sampling_period` is set to
+                1 hour and the length of the buffer is 24, then the data
+                stored in the buffer could correspond to the time range from
+                "2022-01-01 12:00:00" to "2022-01-02 12:00:00" (date chosen
+                arbitrarily here).
+        """
+        self._buffer = buffer
+        self._sampling_period = sampling_period
+        self._time_index_alignment = time_index_alignment
+
+        self._gaps: list[Gap] = []
+        self._datetime_newest = datetime.min
+        self._datetime_oldest = datetime.max
+        self._time_range = (len(self._buffer) - 1) * sampling_period
```

Comment:
Almost all of these are missing type hints. I'm surprised it wasn't caught by the linters.

Reply:
Linters seem to care mostly about things outside of functions (parameters and returns).
Is it our policy to type-hint every single assignment then?

Reply:
I'm pretty sure `mypy` can deduce the types for the fields from the type of the assigned values.  The only place where it can't deduce is for `self._gaps: list[Gap] = []`, because it is an empty list,  and in that case, a type has been provided.